pylutron_caseta/pairing.py

```diff
@@ -0,0 +1,192 @@
+"""Guide the user through pairing and save the necessary files."""
+
+import json
+import logging
+import socket
+import ssl
+import tempfile
+
+from cryptography import x509
+from cryptography.hazmat.backends import default_backend
+from cryptography.hazmat.primitives import hashes, serialization
+from cryptography.hazmat.primitives.asymmetric import rsa
+from cryptography.x509.oid import NameOID  # type: ignore
+
+from .assets import LAP_CA_PEM, LAP_CERT_PEM, LAP_KEY_PEM
+
+LOGGER = logging.getLogger(__name__)
+
+SOCKET_TIMEOUT = 10
+BUTTON_PRESS_TIMEOUT = 180
+CERT_SUBJECT = x509.Name(
+    [x509.NameAttribute(NameOID.COMMON_NAME, "get_lutron_cert.py")]
+)
+
+PAIR_KEY = "key"
+PAIR_CERT = "cert"
+PAIR_CA = "ca"
+PAIR_VERSION = "version"
+
+
+def _generate_private_key():
+    LOGGER.info("Generating a new private key...")
+    return rsa.generate_private_key(
+        public_exponent=65537, key_size=2048, backend=default_backend()
+    )
+
+
+def _convert_private_key_to_pem(private_key):
+    return private_key.private_bytes(
+        encoding=serialization.Encoding.PEM,
+        format=serialization.PrivateFormat.PKCS8,
+        encryption_algorithm=serialization.NoEncryption(),
+    )
+
+
+def _generate_csr(private_key):
+    return (
+        x509.CertificateSigningRequestBuilder()
+        .subject_name(CERT_SUBJECT)
+        .sign(private_key, hashes.SHA256(), default_backend())
+    )
+
+
+class JsonSocket:
+    """A socket that reads and writes json objects."""
+
+    def __init__(self, raw_socket):
+        """Create a JsonSocket wrapping the provided socket."""
+        self._socket = raw_socket
+
+    def read_json(self):
+        """Read an object."""
+        buffer = b""
+        while not buffer.endswith(b"\r\n"):
```

Comment:
Why not use readline? This implementation probably works but is incorrect and may fail. We have a stream-oriented connection to the bridge so we may receive the message terminator and another message in the same read.

Reply:
This is copied code from `get_lutron_cert.py`

Reply:
Replaced with asyncio